add precision parameter to round function

proddata opened the pull request:
## Summary of the changes / Why this improves CrateDB

Enhanced the `round()` scalar function to allow a second input parameter for rounding precision.
When it is specified, the result's type is ``double precision``. Notice that
``round(number)`` and ``round(number, 0)`` return different result types.

```sql
    cr> select round(42.21,1) AS round;
    +-------+
    | round |
    +-------+
    |  42.2 |
    +-------+
    SELECT 1 row in set (... sec)
```

Implementation deviates from PostgreSQL. PostgreSQL always returns `NUMERIC` for `round()`.
Is coherent with `trunc()` scalar function.

closes https://github.com/crate/crate/issues/13964

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

seut commented:
> Implementation deviates from PostgreSQL. PostgreSQL always returns NUMERIC for round().

What is the reason for this deviation?

proddata commented:
> What is the reason for this deviation?

In PostgreSQL: `round` and `trunc` generally returns a `double precision` without the precision parameter and a `numeric` with the precision parameter.
Also the precision parameter is only valid for `numeric` and integer input types

In CrateDB:
- `round`  without the precision parameter returns `integer` or `bigint`
- `trunc`  without the precision parameter returns `integer` or `bigint`
- `trunc`  with the precision parameter returns `double` and allows `double` as input

```sql
> SELECT
pg_typeof(round(1::INT)),
pg_typeof(round(1.0::FLOAT4)),
pg_typeof(round(1.0::FLOAT8)),
pg_typeof(round(1::INT,0)),
pg_typeof(round(1.1::NUMERIC,0));
    pg_typeof     |    pg_typeof     |    pg_typeof     | pg_typeof | pg_typeof 
------------------+------------------+------------------+-----------+-----------
 double precision | double precision | double precision | numeric   | numeric
(1 row)
```

mfussenegger commented on docs/appendices/release-notes/unreleased.rst:
```diff
@@ -70,6 +70,9 @@ Changes
   :ref:`standard_conforming_strings <conf-session-standard_conforming_strings>`
   read-only session setting for improved compatibility with PostgreSQL clients.
 
+- Enhanced the :ref:`round <scalar-round>` scalar function to allow a second
+  input parameter for rounding precision.
+
```
```suggestion
- Added an optional ``precision`` parameter to the :ref:`round <scalar-round>` scalar function.

```

mfussenegger commented on docs/general/builtins/scalar-functions.rst:
```diff
@@ -2201,6 +2204,14 @@ See below for an example::
     +-------+
     SELECT 1 row in set (... sec)
 
+    cr> select round(42.21,1) AS round;
```
```suggestion
    cr> select round(42.21, 1) AS round;
```

mfussenegger commented on server/src/main/java/io/crate/expression/scalar/arithmetic/RoundFunction.java:
```diff
@@ -58,5 +67,41 @@ public static void register(ScalarFunctionModule module) {
+                }
+                double val = n.doubleValue();
+                int numDecimals = nd.intValue();
+                return BigDecimal.valueOf(val).setScale(numDecimals, RoundingMode.HALF_UP).doubleValue();
```
I think if it's not overflowing, `(round(<val> * 10 ^ numDecimals)) / <val> * 10 ^ numDecimals)` would be faster.

proddata replied:
> I think if it's not overflowing, (round(<val> * 10 ^ numDecimals)) / <val> * 10 ^ numDecimals) would be faster.

There is both the problem of overflowing and precision loss as e.g. pointed out [here](https://www.baeldung.com/java-round-decimal-number#mathround-method). While one could check overflow and precision to fit within the max safe integer bounds of a double, the checks become cumbersome and harder to maintain. As this would be primarily be a perform optimisation and I'd expect this function to be mostly used for presentation purposes after calculation on a final result set, I think it isn't worth it. Therefore I would suggest to stay with the BigDecimal approach.

seut replied:
I slightly agree on using `NUMERIC` instead, any objections? @mfussenegger

seut replied:
> > I think if it's not overflowing, (round( * 10 ^ numDecimals)) /  * 10 ^ numDecimals) would be faster.
> 
> There is both the problem of overflowing and precision loss as e.g. pointed out [here](https://www.baeldung.com/java-round-decimal-number#mathround-method). While one could check overflow and precision to fit within the max safe integer bounds of a double, the checks become cumbersome and harder to maintain. As this would be primarily be a perform optimisation and I'd expect this function to be mostly used for presentation purposes after calculation on a final result set, I think it isn't worth it. Therefore I would suggest to stay with the BigDecimal approach.

Is detecting a possible overflow really that complex?
I do not completely follow you argumentation about the actual use-cases of this function as we never know for sure how users are using it. 
But I'm also fine on using your approach now, we can follow up here anytime without a user-facing change

proddata replied:
As mentioned above this method falls apart for inputs like `round(260.775, 2)` with an expected output of `260.78`

```
100 * 260.775 == 26077.499999999996
Math.round(26077.499999999996) == 26077
26077 / 100 == 260.77 but should be 260.78
```

so no. No easy way to detect loss of precision

proddata commented on server/src/main/java/io/crate/expression/scalar/arithmetic/RoundFunction.java:
```diff
@@ -58,5 +67,41 @@ public static void register(ScalarFunctionModule module) {
+            public Number evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Number>... args) {
+                Number n = args[0].value();
+                Number nd = args[1].value();
+                if (null == n || null == nd) {
+                    return null;
+                }
```
ooops ... took that over from`TruncFunction`

surister commented:
This makes some Tableau tests fail.

proddata commented:
> This would be needed for some tests in Tableau

I think the only real open point is the question about the return type. Otherwise this PR is sitting ready for a while now and could be merged.

proddata commented:
I rebased and updated it. As mentioned in the comments, I believe there's no way to avoid using BigDecimal for rounding unless we're okay with deviating from PostgreSQL's behavior and risking loss of precision.

I also prefer to stick with Double as the return type to align with the `trunc` function. CrateDB doesn't support storing NUMERIC types yet, and none of the scalars currently return a numeric value.

matriv reviewed (approved):
To me it looks good. I kind of agree to not return `NUMERIC` (at least now) and we could revisit it once the `NUMERIC` lands in cratedb with full support.

Let's wait for @mfussenegger's opinion as well.

proddata commented:
With NUMERIC storage support now added (https://github.com/crate/crate/pull/16471) should this return NUMERIC instead? 

@matriv / @mfussenegger

mfussenegger reviewed (commented):
Yes I'd tend to return `NUMERIC` for the `round(v, s)` version and leave the existing overload unchanged.

proddata commented on server/src/main/java/io/crate/expression/scalar/arithmetic/RoundFunction.java:
```diff
@@ -62,6 +69,36 @@ public static void register(Functions.Builder module) {
+                }
+                double val = n.doubleValue();
+                int numDecimals = nd.intValue();
+                if(numDecimals < 0) {
```
This is needed to keep the correct scale for `round(123.123,-1)` i.e. `120` instead of `1.2E+2`

mfussenegger commented on server/src/main/java/io/crate/expression/scalar/arithmetic/RoundFunction.java:
```diff
@@ -62,6 +69,36 @@ public static void register(Functions.Builder module) {
+
+            module.add(
+                Signature.builder(NAME, FunctionType.SCALAR)
+                    .argumentTypes(DataTypes.DOUBLE.getTypeSignature(),
+                        DataTypes.INTEGER.getTypeSignature())
```
I wonder if the first argument type should also be `numeric`, or if we should add a second overload to avoid precision loss if the argument is a `numeric` with higher precision.

proddata replied:
I've added a numeric overload and addressed cases involving long values, which could lose precision when cast to double. Could you please take another look?
